File: src/utils/tenant_deletion.py
"""
Tenant deletion utilities.

Provides helper functions to check if a tenant has been marked as deleted.
"""

import asyncpg

from src.utils.logging import get_logger
from src.utils.redis_cache import get_or_compute

logger = get_logger(__name__)

# Redis cache TTL for deleted_at checks (10 minutes)
DELETED_AT_CACHE_TTL_SECONDS = 600


async def is_tenant_deleted(control_db_pool: asyncpg.Pool, tenant_id: str) -> bool:
    """Check if a tenant has been marked as deleted.

    Uses Redis caching with 10-minute TTL to reduce database connection pool pressure.
    Falls back to database if Redis is unavailable.

    Args:
        control_db_pool: Control database connection pool
        tenant_id: The tenant ID to check

    Returns:
        bool: True if tenant is deleted (deleted_at is not null), False otherwise
    """
    cache_key = f"tenant:deleted:{tenant_id}"

    async def fetch_from_db() -> bool:
        """Query database for tenant deletion status."""
        async with control_db_pool.acquire() as conn:
            deleted_at = await conn.fetchval(
                "SELECT deleted_at FROM tenants WHERE id = $1", tenant_id
            )

        is_deleted = deleted_at is not None

        if is_deleted:
            logger.warning(f"Tenant {tenant_id} is marked as deleted (deleted at {deleted_at})")

        return is_deleted

    # Use cache helper to get or compute the value
    return await get_or_compute(
        cache_key=cache_key,
        compute_fn=fetch_from_db,
        serialize_fn=lambda val: "1" if val else "0",
        deserialize_fn=lambda s: s == "1",
        ttl_seconds=DELETED_AT_CACHE_TTL_SECONDS,
    )

File: js-services/slack-bot/src/utils/tenantDeletion.ts
/**
 * Tenant deletion utility functions
 * Provides utilities to check if a tenant has been marked as deleted
 */

import { Pool } from 'pg';
import { logger } from './logger';
import { getOrCompute } from '@corporate-context/backend-common';

let controlDbPool: Pool | null = null;

// Redis cache TTL for deleted_at checks (10 minutes)
const DELETED_AT_CACHE_TTL_SECONDS = 600;

/**
 * Get or create the control database connection pool
 */
function getControlDbPool(): Pool | null {
  if (controlDbPool) {
    return controlDbPool;
  }

  const databaseUrl = process.env.CONTROL_DATABASE_URL;
  if (!databaseUrl) {
    logger.error(
      'Control database URL not configured. Set CONTROL_DATABASE_URL environment variable.'
    );
    return null;
  }

  const isLocalDb = databaseUrl.includes('localhost') || databaseUrl.includes('127.0.0.1');
  controlDbPool = new Pool({
    connectionString: databaseUrl,
    ssl: isLocalDb ? false : { rejectUnauthorized: false },
    max: 5,
    min: 1,
    connectionTimeoutMillis: 30000,
    idleTimeoutMillis: 30000,
    maxUses: 7500,
  });

  return controlDbPool;
}

/**
 * Check if a tenant is deleted
 * Uses Redis caching with 10-minute TTL to reduce database connection pool pressure.
 * Falls back to database if Redis is unavailable.
 *
 * @param tenantId - The tenant ID to check
 * @returns Promise<boolean> - True if tenant is deleted (deleted_at is not null), false otherwise
 */
export async function isTenantDeleted(tenantId: string): Promise<boolean> {
  const cacheKey = `tenant:deleted:${tenantId}`;

  const fetchFromDb = async (): Promise<boolean> => {
    const pool = getControlDbPool();
    if (!pool) {
      logger.error('Control database not available for tenant deletion check');
      return false;
    }

    try {
      const result = await pool.query('SELECT deleted_at FROM tenants WHERE id = $1', [tenantId]);

      if (result.rows.length === 0) {
        logger.warn(`Tenant ${tenantId} not found in control database`);
        return false;
      }

      const deletedAt = result.rows[0].deleted_at;
      const isDeleted = deletedAt !== null;

      if (isDeleted) {
        logger.warn(`Tenant ${tenantId} is marked as deleted (deleted at ${deletedAt})`);
      }

      return isDeleted;
    } catch (error) {
      logger.error(`Error checking tenant deletion status for ${tenantId}`, { error });
      return false;
    }
  };

  // Use cache helper to get or compute the value
  return await getOrCompute(
    cacheKey,
    fetchFromDb,
    (val) => (val ? '1' : '0'),
    (s) => s === '1',
    DELETED_AT_CACHE_TTL_SECONDS
  );
}

/**
 * Close the control database connection pool
 */
export async function closeControlDbPool(): Promise<void> {
  if (controlDbPool) {
    await controlDbPool.end();
    controlDbPool = null;
  }
}
